src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java

```diff
@@ -539,12 +549,51 @@ public long importKey(long hSession, CK_ATTRIBUTE[] attributes) throws PKCS11Exc
             if (attr.type == CKA_KEY_TYPE) {
                 keyType = attr.getLong();
             }
-            if (attr.type == CKA_VALUE) {
-                keyBytes = attr.getByteArray();
+            ckAttrsMap.put(attr.type, attr);
+        }
+
+        if(keyClass == CKO_PRIVATE_KEY) {
+            if(keyType == CKK_RSA) {
+                try {
+                    keyBytes = RSAPrivateCrtKeyImpl.newKey(
+                        KeyType.RSA,
+                        null,
+                        ckAttrsMap.getOrDefault(CKA_MODULUS, new CK_ATTRIBUTE(CKA_MODULUS, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PUBLIC_EXPONENT, new CK_ATTRIBUTE(CKA_PUBLIC_EXPONENT, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PRIVATE_EXPONENT, new CK_ATTRIBUTE(CKA_PRIVATE_EXPONENT, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PRIME_1, new CK_ATTRIBUTE(CKA_PRIME_1, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_PRIME_2, new CK_ATTRIBUTE(CKA_PRIME_2, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_EXPONENT_1, new CK_ATTRIBUTE(CKA_EXPONENT_1, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_EXPONENT_2, new CK_ATTRIBUTE(CKA_EXPONENT_2, BigInteger.ZERO)).getBigInteger(),
+                        ckAttrsMap.getOrDefault(CKA_COEFFICIENT, new CK_ATTRIBUTE(CKA_COEFFICIENT, BigInteger.ZERO)).getBigInteger()
+                    ).getEncoded();
+                } catch (InvalidKeyException e) {
+                    throw new PKCS11Exception(0x00000005L, null);
+                }
+            } else if (keyType == CKK_EC) {
+                try {
+                    keyBytes = ECUtil.generateECPrivateKey(
+                        ckAttrsMap.getOrDefault(CKA_VALUE, new CK_ATTRIBUTE(CKA_VALUE, BigInteger.ZERO)).getBigInteger(),
+                        ECUtil.getECParameterSpec(
+                            Security.getProvider("SunEC"),
+                            ckAttrsMap.get(CKA_EC_PARAMS).getByteArray())
+                    ).getEncoded();
+                    // If key is private and of EC type, NSS may require CKA_NETSCAPE_DB
+                    // attribute to unwrap it. Otherwise, C_UnwrapKey will produce an Exception.
+                    if (token.config.getNssNetscapeDbWorkaround() && !ckAttrsMap.containsKey(CKA_NETSCAPE_DB)) {
+                        ckAttrsMap.put(CKA_NETSCAPE_DB, new CK_ATTRIBUTE(CKA_NETSCAPE_DB, BigInteger.ZERO));
+                    }
+                } catch (IOException | InvalidKeySpecException e) {
+                    throw new PKCS11Exception(0x00000005L, null);
+                }
             }
+        } else if (keyClass == CKO_SECRET_KEY) {
+            keyBytes = ckAttrsMap.get(CKA_VALUE).getByteArray();
```

Comment:
What guarantees that `CKA_VALUE` exists?

Reply:
As far as I know, an Generic Secret Key Object Attributes contains both CKA_VALUE and CKA_VALUE_LEN. The CKA_VALUE attribute of a generic secret key must be provided when the secret key object is created.

Reply:
I think this should continue to avoid making the assumption that `CKA_VALUE` exists in the list of attributes.
```java
            CK_ATTRIBUTE value = ckAttrsMap.get(CKA_VALUE);
            if (value != null) {
                keyBytes = value.getByteArray();
            }
```